src/rust/lib_ccxr/src/util/time/units.rs

```diff
@@ -0,0 +1,634 @@
+    /// let timestamp = Timestamp::from_millis(6524365);
+    /// assert_eq!(timestamp.to_hms_millis_time(':').unwrap(), "01:48:44:365");
+    /// ```
+    pub fn to_hms_millis_time(&self, sep: char) -> Result<String, TimestampError> {
+        let mut s = String::new();
+        self.write_hms_millis_time(&mut s, sep)?;
+        Ok(s)
+    }
+
+    /// Returns a formatted [`Timestamp`] using ctime's format.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use lib_ccxr::util::time::Timestamp;
+    /// let timestamp = Timestamp::from_millis(6524365);
+    /// assert_eq!(timestamp.to_ctime().unwrap(), "Thu Jan 01 01:48:44 1970");
+    /// ```
+    pub fn to_ctime(&self) -> Result<String, TimestampError> {
+        let mut s = String::new();
+        self.write_ctime(&mut s)?;
+        Ok(s)
+    }
+
+    /// Returns a formatted [`Timestamp`] using format specified by [`TimestampFormat`].
+    ///
+    /// See [`TimestampFormat`] for examples.
+    pub fn to_formatted_time(&self, format: TimestampFormat) -> Result<String, TimestampError> {
+        let mut s = String::new();
+        self.write_formatted_time(&mut s, format)?;
+        Ok(s)
+    }
+
+    /// Creates a [`Timestamp`] by parsing `input` using format `SS` or `MM:SS` or `HH:MM:SS`.
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use lib_ccxr::util::time::Timestamp;
+    /// let timestamp = Timestamp::parse_optional_hhmmss_from_str("01:12:45").unwrap();
+    /// assert_eq!(timestamp, Timestamp::from_millis(4_365_000));
+    /// ```
+    pub fn parse_optional_hhmmss_from_str(input: &str) -> Result<Timestamp, TimestampError> {
+        let mut numbers = input
+            .split(':')
+            .map(|x| x.parse::<u8>().map_err(|_| TimestampError::ParsingError))
+            .rev();
+
+        let mut millis: u64 = 0;
+
+        let seconds: u64 = numbers.next().ok_or(TimestampError::ParsingError)??.into();
+        if seconds > 59 {
+            return Err(TimestampError::InputOutOfRangeError);
+        }
+        millis += seconds * 1000;
+
+        if let Some(x) = numbers.next() {
+            let minutes: u64 = x?.into();
+            if minutes > 59 {
+                return Err(TimestampError::InputOutOfRangeError);
+            }
+            millis += 60_000 * minutes;
+        }
+
+        if let Some(x) = numbers.next() {
+            let hours: u64 = x?.into();
+            millis += 3_600_000 * hours;
+        }
+
+        if numbers.next().is_some() {
+            return Err(TimestampError::ParsingError);
+        }
+
+        Ok(Timestamp::from_millis(millis.try_into()?))
+    }
+}
+
+/// Represent the number of clock ticks as defined in Mpeg standard.
+///
+/// This number can never be negetive.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd, Add, Sub)]
+pub struct MpegClockTick(i64);
+
+impl MpegClockTick {
+    /// The ratio to convert a clock tick to time duration.
+    pub const MPEG_CLOCK_FREQ: i64 = 90000;
```

Comment:
This is a global value in C. AFAIK, If you redefine it here it will contain an incorrect value, if the C constant is changed

Reply:
I have converted it into a AtomicI64

Reply:
Why is there a need to use atomics here? We are running a single thread

Reply:
The only way to create a global mutable in rust is to declare a static variable. Because a static variable can be changed by multiple threads, it requires atomics. There is a way to bypass it using a mutable static variable and using unsafe eveytime we try to access or change it. Do you want me go with this approach? I was hesitant at first since it used unsafe rust.

Reply:
Yes, as we don't use any multithreading, I would prefer if you just re-use the C global variable and not redeclare it in rust.
This is what we have decided in #1547 too

Reply:
I have now re-used `MPEG_CLOCK_FREQ` from C.